packages/toolpad-app/src/utils/useSearch.tsx

```diff
@@ -0,0 +1,40 @@
+import * as React from 'react';
+import { TextField } from '@mui/material';
+import Search from '@mui/icons-material/SearchOutlined';
+
+/**
+ * Creates a search field.
+ * @param elements The elements to search through.
+ * @param  {string=} placeholder The value of the placeholder text
+ * @returns The search field and the filtered elements
+ */
+
+function useSearch<T>(elements: T[], placeholder?: string): [JSX.Element, T[]] {
+  const [searchText, setSearchText] = React.useState('');
+
+  const handleSearchInput = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
+    setSearchText(event.target.value);
+  }, []);
+
+  const filteredElements = React.useMemo(() => {
+    return elements.filter((elem: any) =>
+      elem.name.toLowerCase().includes(searchText.toLowerCase()),
+    );
+  }, [elements, searchText]);
+
+  return [
+    <TextField
```

Comment:
This pattern of returning both react component and filtered values as a pair looks pretty uncommon. Since the logic is so trivial, I'd personally consider keeping component itself inside the main component.

To separate filtering/searching logic, maybe we could keep this `useSearch` hook, but instead of returning component we would return only filtered results. Then the input for hook would be `allApps, searchValue`:
```
const [searchValue, setSearchValue] = useState('')
const filteredApps = useSearch(apps, searchValue)
...
<TextField key='search' value={searchValue} onChange={(event) => setSearchValue(event.target.value)}
```

Reply:
Don't create hooks until a clear pattern of reuse emerges. The logic can go straight into `toolpad/Apps/index.ts`. We will extract reusable patterns once we align implementation between apps and connections